csrc/scheduler/matmul_utils.cpp

```diff
@@ -294,6 +315,161 @@ std::string isMatmulFusionDefinitionSupported(
+    const std::vector<int64_t>& sizes,
+    const std::vector<int64_t>& strides) {
+  int64_t vec_size = scheduler_utils::maxVectorizationWidth(data_ptr_int);
+  vec_size = std::min(vec_size, 16l);
+  vec_size /= dataTypeSize(tv->dtype());
+  vec_size = std::max(vec_size, 1l);
+  if (vec_size == 1l) {
+    return vec_size;
+  }
+
+  NVF_ERROR(sizes.size() == strides.size());
+  NVF_ERROR((int64_t)sizes.size() == tv->nDims());
+
+  std::vector<std::optional<bool>> noreductions_contig;
+  for (size_t i : c10::irange(tv->getMaybeAllocationDomain().size())) {
+    if (tv->getMaybeAllocationDomain()[i]->isReduction()) {
+      continue;
+    }
+    noreductions_contig.push_back(tv->getContiguity().at(i));
+  }
+
+  bool inner_contiguous = false;
+  for (size_t i : c10::irange(sizes.size())) {
+    int64_t stride = strides.at(i);
+    int64_t size = sizes.at(i);
+    if (stride == 1) {
+      std::optional<bool> contig = noreductions_contig.at(i);
+      if (contig.has_value() && !contig.value()) {
+        // If TensorView is marked discontiguous in inner dimension, we cannot
+        // vectorize regardless of input.
+        return 1l;
+      }
+      if (inner_contiguous) {
+        // There was already a contiguous inner dimension, so that means there
+        // are multiple stride==1 dimensions in this tensor. This means that
+        // there will be overlapping rows offset from one another by a single
+        // element. In such cases we cannot vectorize. An exception is when an
+        // outer dimension has both stride==1 and size==1, in which case we
+        // could ignore its stride. Currently we ignore that case and just
+        // return 1 whenever we detect multiple stride==1 dimensions.
+        return 1l;
+      }
+      inner_contiguous = true;
+      vec_size =
+          std::min(vec_size, scheduler_utils::maxVectorizationWidth(size));
+    } else {
+      vec_size =
+          std::min(vec_size, scheduler_utils::maxVectorizationWidth(stride));
+    }
+  }
+
+  if (!inner_contiguous) {
+    // Tensor is discontiguous
+    return 1l;
+  }
+
+  return vec_size;
+}
+
+MatmulParams::SupportedVectorization getSupportedVectorization(
+    const mma_utils::RolesMap& roles_map,
+    SchedulerRuntimeInfo& runtime_info) {
+  auto getMinVectorization = [&roles_map,
+                              &runtime_info](MatmulRole role) -> int64_t {
+    int64_t vec_size = 16; // max vectorization size
+    const auto it = roles_map.find(role);
+    if (it == roles_map.end()) {
+      return 16;
+    }
+    for (TensorView* tv : it->second) {
+      // TODO: handle the case when tv is not a Fusion input by filling default
+      // contiguous strides based on tv->getMaybeAllocationDomain() and data
+      // pointer aligned to 16 bytes.
+
+      const std::vector<IterDomain*>& alloc = tv->getMaybeAllocationDomain();
+
+      // Get sizes for non-reduction dims in allocation domain
+      std::vector<int64_t> sizes;
```

Comment:
Could we also move the size to `runtime_info` to share the permutation code of stride?

Reply:
Makes sense. I was just using expression evaluator to get that but it's more compact to just load it directly.